### PR TITLE
[1LP][RFR] Catalog Widgetastic Conversion

### PR DESCRIPTION
**Pull request by sshveta**
{{pytest: -v --long-running cfme/tests/services/test_catalog.py }}

### CODE REVIEW

File: cfme/services/catalogs/catalog.py [hunk 1]
```diff
@@ -1,115 +1,104 @@
...
 
-def _all_catalogs_add_new(_):
-    if BZ(1213863).blocks:
-        sel.pytest.skip("Blocker on the bug 1213863")
-    catalog_tree('All Catalogs')
-    cfg_btn('Add a New Catalog')
+    @property
+    def is_displayed(self):
+        return (
+            super(CatalogsView, self).is_displayed and
```
Comment (mfalesni) [inline cfme/services/catalogs/catalog.py]: **REQUIRED**: Do not use super() like this. ServicesCatalogView should check that it is in ServicesCatalog and it should also check that accordion is enabled and such things - that it actually is usable for the steps to happen since that ensures correct matching of locations. I strongly advocate for creating a property, let's say `in_servies_catalogs` and use that one instead in the subclasses.

Then the parent class would look like.

```python
@property
def in_services_catalogs(self):
    return 'logged in as current user and correct menu item selected'

@property
def is_displayed(self):
    return self.in_services_catalogs and not accordion.is_dimmed and 'whatever else to ensure that the view is correct'
```

Then in the child class, like this one, you will just replace this line with

```python
self.in_services_catalogs and  # ...
```

You might see the super() usage in some places, yes, somewhere it makes sense and in the other places I am going to remove them :)

File: cfme/services/catalogs/catalog.py [hunk 2]
```diff
@@ -1,115 +1,104 @@
...
 
-item_form = web_ui.Form(
-    fields=[('type_select', "//select[@id='st_prov_type']"),
-            ('name_text', Input("name")),
-            ('description_text', Input("description")),
-            ('display_checkbox', Input("display")),
-            ('add_button', form_buttons.add)])
+    add_button = Button('Add')
+    save_button = Button('Save')
+    cancel_button = Button('Cancel')
```
Comment (mfalesni) [inline cfme/services/catalogs/catalog.py]: **REQUIRED**: The CatalogForm child classes also specify the buttons. Currently it is customary to have them in the child classes only (for the different between add/save). So you should remove them here.

File: cfme/services/catalogs/catalog.py [hunk 3]
```diff
@@ -1,115 +1,104 @@
...
+        view.fill({'name_text': updates.get('name', None),
+                   'description_text': updates.get('description', None),
+                   'button_multiselect': updates.get('items', None)})
+        view.save_button.click()
+        view.flash.assert_success_message('Catalog "{}" was saved'.format(self.name))
+
+    def delete(self):
+        view = navigate_to(self, "Details")
+        view.configuration.item_select(version.pick({version.LOWEST: 'Remove Item from the VMDB',
+                    '5.7': 'Remove Catalog'}), handle_alert=True)
```
Comment (mfalesni) [inline cfme/services/catalogs/catalog.py]: **REQUIRED**: Indenting

```python
view.configuration.item_select(
    version.pick({
        version.LOWEST: 'Remove Item from the VMDB',
        '5.7': 'Remove Catalog'}),
    handle_alert=True)
```

File: cfme/services/catalogs/catalog.py [hunk 4]
```diff
@@ -126,56 +115,38 @@ def exists(self):
...
-        if paginator.page_controls_exist():
-            # Make sure nothing is selected
-            sel.check(paginator.check_all())
-            sel.uncheck(paginator.check_all())
+        self.view.catalogs.tree.click_path("All Catalogs")
 
 
-@navigator.register(Catalog, 'Add')
-class Add(CFMENavigateStep):
+@navigator.register(Catalog, 'AddCatalog')
```
Comment (mfalesni) [inline cfme/services/catalogs/catalog.py]: Why the change? Convention says that we should use Add/Edit/Details and such names, so you never have to look it up. If you have multiple classes defined in one module and the Add steps' classes' names clash, you can rename the class but keep the step name as Add ...

File: cfme/services/catalogs/catalog.py [hunk 5]
```diff
@@ -126,56 +115,38 @@ def exists(self):
...
-        return match_page(summary='Catalog "{}"'.format(self.obj.name))
+    VIEW = DetailsCatalogView
 
     def step(self):
-        # Best effort to make sure the catalog is in the list
-        if paginator.page_controls_exist():
-            paginator.results_per_page(1000)
-        listview_table.click_row_by_cells(
-            {'Name': self.obj.name, 'Description': self.obj.description})
+        self.view.catalogs.tree.click_path("All Catalogs", self.obj.name)
```
Comment (mfalesni) [inline cfme/services/catalogs/catalog.py]: **REQUIRED**: Use `self.prerequisite_view` since to get to the location, you usually need the previous view and do something with it to get to the location this step represents. In this case it probably works because the views are inherited, but this leads to bad copypasting and it might also lead to pulled hair when something changes :)

File: cfme/services/catalogs/catalog.py [hunk 6]
```diff
@@ -126,56 +115,38 @@ def exists(self):
...
+class EditCatalog(CFMENavigateStep):
     prerequisite = NavigateToSibling('Details')
 
-    def am_i_here(self):
-        match_page(summary='Editing Catalog "{}"'.format(self.obj.name))
+    VIEW = EditCatalogView
 
     def step(self):
-        cfg_btn('Edit this Item')
+        self.view.configuration.item_select('Edit this Item')
```
Comment (mfalesni) [inline cfme/services/catalogs/catalog.py]: **REQUIRED**: Same with the prerequisite view as above.

**Comment by sshveta** [comment]
Done , Done and Done , Please review again

File: cfme/services/catalogs/catalog.py [hunk 7]
```diff
@@ -1,115 +1,106 @@
...
-        # For element not found exception (To be removed)
-        sel.sleep(5)
-        web_ui.fill(form, {'name_text': self.name,
-                           'description_text': self.description,
-                           'button_multiselect': self.items},
-                    action=form.add_button)
+        view = navigate_to(self, 'Add')
+        view.fill({'name_text': self.name,
+                   'description_text': self.description,
+                   'button_multiselect': self.items})
```
Comment (izapolsk) [inline cfme/services/catalogs/catalog.py]: REQUIRED: I haven't found 'button_multiselect' among view or form controls. Are you sure you haven't missed it ?

File: cfme/services/catalogs/catalog.py [hunk 8]
```diff
@@ -1,115 +1,106 @@
...
-        navigate_to(self, 'Add')
-        # For element not found exception (To be removed)
-        sel.sleep(5)
-        web_ui.fill(form, {'name_text': self.name,
-                           'description_text': self.description,
-                           'button_multiselect': self.items},
-                    action=form.add_button)
+        view = navigate_to(self, 'Add')
+        view.fill({'name_text': self.name,
+                   'description_text': self.description,
```
Comment (izapolsk) [inline cfme/services/catalogs/catalog.py]: OPTIONAL: it make sense to remove text from control names.

File: cfme/services/catalogs/catalog.py [hunk 9]
```diff
@@ -1,115 +1,106 @@
...
+                   'button_multiselect': self.items})
+        view.add_button.click()
         flash_str = 'Catalog "{}" was saved'
-        flash.assert_success_message(flash_str.format(self.name))
+        view.flash.assert_success_message(flash_str.format(self.name))
```
Comment (izapolsk) [inline cfme/services/catalogs/catalog.py]: OPTIONAL: I'd prefer to see formatted string and format statement together. readability counts.

File: cfme/services/catalogs/catalog.py [hunk 10]
```diff
@@ -1,115 +1,106 @@
...
-        elif from_dest == 'Details':
-            cfg_btn(version.pick({version.LOWEST: 'Remove Item from the VMDB',
-                    '5.7': 'Remove Catalog'}), invokes_alert=True)
-
-        sel.handle_alert()
-        flash.assert_success_message(
+        view = navigate_to(self, 'Edit')
+        view.fill({'name_text': updates.get('name', None),
+                   'description_text': updates.get('description', None),
+                   'button_multiselect': updates.get('items', None)})
```
Comment (izapolsk) [inline cfme/services/catalogs/catalog.py]: REQUIRED: ^

File: cfme/services/catalogs/catalog.py [hunk 11]
```diff
@@ -126,56 +117,38 @@ def exists(self):
 class All(CFMENavigateStep):
     prerequisite = NavigateToAttribute('appliance.server', 'LoggedIn')
 
-    def am_i_here(self):
-        return match_page(summary='All Catalogs')
+    VIEW = CatalogsView
```
Comment (izapolsk) [inline cfme/services/catalogs/catalog.py]: REQUIRED: could you please put VIEW before prerequisite ?

File: cfme/services/catalogs/catalog.py [hunk 12]
```diff
@@ -126,56 +117,38 @@ def exists(self):
...
+        self.view.catalogs.tree.click_path("All Catalogs")
 
 
 @navigator.register(Catalog, 'Add')
 class Add(CFMENavigateStep):
     prerequisite = NavigateToSibling('All')
 
-    def am_i_here(self):
-        return match_page(summary='Adding a new Catalog')
+    VIEW = AddCatalogView
```
Comment (izapolsk) [inline cfme/services/catalogs/catalog.py]: ^

File: cfme/services/catalogs/catalog.py [hunk 13]
```diff
@@ -126,56 +117,38 @@ def exists(self):
...
 
 
 @navigator.register(Catalog, 'Details')
 class Details(CFMENavigateStep):
     prerequisite = NavigateToSibling('All')
 
-    def am_i_here(self):
-        # header text as separate var to reduce quote escaping
-        return match_page(summary='Catalog "{}"'.format(self.obj.name))
+    VIEW = DetailsCatalogView
```
Comment (izapolsk) [inline cfme/services/catalogs/catalog.py]: ^

File: cfme/services/catalogs/catalog.py [hunk 14]
```diff
@@ -126,56 +117,38 @@ def exists(self):
...
+        self.prerequisite_view.catalogs.tree.click_path("All Catalogs", self.obj.name)
 
 
 @navigator.register(Catalog, 'Edit')
 class Edit(CFMENavigateStep):
     prerequisite = NavigateToSibling('Details')
 
-    def am_i_here(self):
-        match_page(summary='Editing Catalog "{}"'.format(self.obj.name))
+    VIEW = EditCatalogView
```
Comment (izapolsk) [inline cfme/services/catalogs/catalog.py]: ^

File: cfme/services/catalogs/catalog.py (hunk 7 above)
Comment (sshveta) [inline cfme/services/catalogs/catalog.py]: removing multiselect , not required in making catalog

File: cfme/services/catalogs/catalog.py (hunk 9 above)
Comment (sshveta) [inline cfme/services/catalogs/catalog.py]: Done

File: cfme/services/catalogs/catalog.py (hunk 8 above)
Comment (sshveta) [inline cfme/services/catalogs/catalog.py]: Done

File: cfme/services/catalogs/catalog.py (hunk 11 above)
Comment (sshveta) [inline cfme/services/catalogs/catalog.py]: Got it

File: cfme/services/catalogs/catalog.py [hunk 15]
```diff
@@ -1,115 +1,103 @@
...
-            cfg_btn(version.pick({version.LOWEST: 'Remove Item from the VMDB',
-                    '5.7': 'Remove Catalog'}), invokes_alert=True)
-
-        sel.handle_alert()
-        flash.assert_success_message(
+        view = navigate_to(self, 'Edit')
+        view.fill({'name': updates.get('name', None),
+                   'description': updates.get('description', None)})
+        view.save_button.click()
+        view.flash.assert_success_message('Catalog "{}" was saved'.format(self.name))
```
Comment (mfalesni) [inline cfme/services/catalogs/catalog.py]: Check this out: https://github.com/ManageIQ/integration_tests/blob/master/cfme/automate/explorer/namespace.py#L217

This is how I rewrote the CRUD methods for use with widgetastic.

File: cfme/services/catalogs/catalog.py [hunk 16]
```diff
@@ -1,115 +1,103 @@
...
-                    '5.7': 'Remove Catalogs'}), invokes_alert=True)
-        elif from_dest == 'Details':
-            cfg_btn(version.pick({version.LOWEST: 'Remove Item from the VMDB',
-                    '5.7': 'Remove Catalog'}), invokes_alert=True)
-
-        sel.handle_alert()
-        flash.assert_success_message(
+        view = navigate_to(self, 'Edit')
+        view.fill({'name': updates.get('name', None),
+                   'description': updates.get('description', None)})
```
Comment (mfalesni) [inline cfme/services/catalogs/catalog.py]: No need to tinker with the updates dict, just throw it in fill :)

File: cfme/services/catalogs/catalog.py [hunk 17]
```diff
@@ -1,115 +1,110 @@
...
+            self.catalogs.is_opened
+        )
 
 
 class Catalog(Updateable, Pretty, Navigatable):
-    """Represents a Catalog"""
-    pretty_attrs = ['name', 'items']
 
     def __init__(self, name=None, description=None, items=None, appliance=None):
+        Navigatable.__init__(self, appliance)
```
Comment (mfalesni) [inline cfme/services/catalogs/catalog.py]: Can we keep the keyword reference here? `Navigatable.__init__(self, appliance=appliance)`

File: cfme/services/catalogs/catalog.py [hunk 18]
```diff
@@ -1,115 +1,110 @@
...
-                           'description_text': self.description,
-                           'button_multiselect': self.items},
-                    action=form.add_button)
-        flash_str = 'Catalog "{}" was saved'
-        flash.assert_success_message(flash_str.format(self.name))
+        view = navigate_to(self, 'Add')
+        view.fill({'name': self.name,
+                   'description': self.description})
+        view.add_button.click()
+        view.flash.assert_success_message('Catalog "{}" was saved'.format(self.name))
```
Comment (mfalesni) [inline cfme/services/catalogs/catalog.py]: So, it should actually create a view that it is supposed to have landed on and then assert is_displayed on it. Then use that view for checking flash. Something like:

```python
view.fill(...)
view.add_button.click()
view = self.create_view(ViewItShouldHaveLandedOn)  # In case of update(self, updates) you should also pass override=updates
assert view.is_displayed
view.flash.assert_no_error()
view.flash.assert_success_message('Catalog "{}" was saved'.format(self.name))
```

create_view passes self automatically in the view as the context object.

From the latest widgetastic version you can use fill_with which triggers the appropriate actions upon filling:
```python
if view.fill_with({...}, on_change=view.add_button, no_change=view.cancel_button):
    view = self.create_view(ViewItShouldHaveLandedOn)  # In case of update(self, updates) you should also pass override=updates
    assert view.is_displayed
    view.flash.assert_no_error()
    view.flash.assert_success_message('Catalog "{}" was saved'.format(self.name))
else:  # when cancel
    view = self.create_view(AnotherViewItShouldHaveLandedOn)  # In case of update(self, updates) DO NOT pass updates here
    assert view.is_displayed
    view.flash.assert_no_error()
    view.flash.assert_success_message('Add was cancelled')
```

@psav is this the way how we should go forward? More checking? I do it like this with automate.

Also the self.name, in some places the display names, descriptions and such things change the used string based on what is and what is not specified. I think that since this is a conversion it should not be the case but just reminding :)

Comment (psav) [inline cfme/services/catalogs/catalog.py]: Yeh I think this is right :) Makes me think we could have a special fn that is used to assert actions as there is some DRY inside that, but I'm happy with what you have suggested, it looks like what I would expect. The only confusing thing is the reuse of the view variable.

Comment (mfalesni) [inline cfme/services/catalogs/catalog.py]: Well, you know, only one view can be valid at the time so if you consider the `view` variable as something you are working on and supposed to be there ...

Comment (sshveta) [inline cfme/services/catalogs/catalog.py]: Did this for Edit and delete but cannot do it for Add method as the view is different when I try to add a duplicate catalog in a test.

Comment (sshveta) [inline cfme/services/catalogs/catalog.py]: I have done assert flash message before assert view in Add method so that while creating duplicate catalog item I can catch the error and view does not conflict . Please see if its okay.

File: cfme/services/catalogs/catalog.py [hunk 19]
```diff
@@ -1,115 +1,121 @@
...
-    fields=[('type_select', "//select[@id='st_prov_type']"),
-            ('name_text', Input("name")),
-            ('description_text', Input("description")),
-            ('display_checkbox', Input("display")),
-            ('add_button', form_buttons.add)])
+    @property
+    def is_displayed(self):
+        return (
+            self.title.text == "Adding a new Catalog" and
+            self.catalogs.is_opened
```
Comment (mfalesni) [inline cfme/services/catalogs/catalog.py]: so, it should also check in_explorer. And the is_opened should come before the title text. So the order then would be in_explorer, catalogs, title text.

File: cfme/services/catalogs/catalog.py [hunk 20]
```diff
@@ -1,115 +1,121 @@
...
-    catalog_tree('All Catalogs')
-    cfg_btn('Add a New Catalog')
+    save_button = Button('Save')
+    reset_button = Button('Reset')
+
+    @property
+    def is_displayed(self):
+        return (
+            self.title.text == "Editing Catalog {}".format(self.name) and
+            self.catalogs.is_opened
```
Comment (mfalesni) [inline cfme/services/catalogs/catalog.py]: Same as my previous comment.
